File: clojure/src/circle_db/constructs.clj
(ns circle-db.constructs)

(defrecord Datom [entity-id attr-name value])

(defrecord Attribute [name value type cardinality ts prev-ts])

(defrecord Entity [id attrs])

(defrecord Layer [storage eavt avet veat vaet])

(defrecord Database [layers top-id curr-time])

(defn make-attr [name value type & {:keys [cardinality] :or {cardinality :db/single}}]
  (->Attribute name value type cardinality -1 -1))

(defn make-entity
  ([]    (->Entity :db/no-id-yet {}))
  ([id]  (->Entity id {})))

(defn make-layer []
  (->Layer {} {} {} {} {}))

(defn make-db []
  (->Database [(make-layer)] 0 0))

File: python/src/constructs.py
from dataclasses import dataclass, field
from typing import Any


@dataclass
class Datom:
    entity_id: int
    attr_name: str
    value: Any


@dataclass
class Database:
    layers: list
    top_id: int
    curr_time: int


def make_db() -> "Database":
    return Database(layers=[Layer()], top_id=0, curr_time=0)


@dataclass
class Layer:
    storage: dict = field(default_factory=dict)
    eavt: dict = field(default_factory=dict)
    avet: dict = field(default_factory=dict)
    veat: dict = field(default_factory=dict)
    vaet: dict = field(default_factory=dict)


@dataclass
class Entity:
    id: int | str  # str used as sentinel ":db/no-id-yet" before ID is assigned
    attrs: dict = field(default_factory=dict)


@dataclass
class Attribute:
    name: str
    value: Any
    type: str
    cardinality: str
    ts: int = -1
    prev_ts: int = -1
